// @prismatic-io/spectral@9.0.4/dist/index.js
"use strict";
/**
 * This module contains functions to help custom component
 * authors create inputs, actions, and components that can
 * be processed by the Prismatic API.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.testing = exports.util = exports.componentManifests = exports.oauth2Connection = exports.onPremConnection = exports.connection = exports.input = exports.dataSource = exports.trigger = exports.action = exports.component = exports.componentManifest = exports.connectionConfigVar = exports.dataSourceConfigVar = exports.configVar = exports.configPage = exports.flow = exports.integration = void 0;
const convertComponent_1 = require("./serverTypes/convertComponent");
const convertIntegration_1 = require("./serverTypes/convertIntegration");
/**
 * This function creates a Integration object that can be
 * imported into the Prismatic API. For information on using
 * this function to write code native integrations, see
 * https://prismatic.io/docs/code-native-integrations/.
 * @param definition An IntegrationDefinition type object.
 * @returns This function returns an integration object that has the shape the Prismatic API expects.
 */
const integration = (definition) => {
    var _a;
    const integrationDefinition = (0, convertIntegration_1.convertIntegration)(definition);
    if (((_a = process.env) === null || _a === void 0 ? void 0 : _a.DEBUG) === "true") {
        console.info(integrationDefinition.codeNativeIntegrationYAML);
    }
    return integrationDefinition;
};
exports.integration = integration;
/**
 * For information on writing Code Native Integrations, see
 * https://prismatic.io/docs/code-native-integrations/#adding-flows.
 * @param definition A Flow type object.
 * @returns This function returns a flow object that has the shape the Prismatic API expects.
 */
const flow = (definition) => definition;
exports.flow = flow;
/**
 * For information on writing Code Native Integrations, see
 * https://prismatic.io/docs/code-native-integrations/#adding-config-pages.
 * @param definition A Config Page type object.
 * @returns This function returns a config page object that has the shape the Prismatic API expects.
 */
const configPage = (definition) => definition;
exports.configPage = configPage;
/**
 * For information on writing Code Native Integrations, see
 * https://prismatic.io/docs/code-native-integrations/#adding-config-vars.
 * @param definition A Config Var type object.
 * @returns This function returns a standard config var object that has the shape the Prismatic API expects.
 */
const configVar = (definition) => definition;
exports.configVar = configVar;
/**
 * For information on writing Code Native Integrations, see
 * https://prismatic.io/docs/code-native-integrations/#adding-config-vars.
 * @param definition A Data Source Config Var type object.
 * @returns This function returns a data source config var object that has the shape the Prismatic API expects.
 */
const dataSourceConfigVar = (definition) => definition;
exports.dataSourceConfigVar = dataSourceConfigVar;
/**
 * For information on writing Code Native Integrations, see
 * https://prismatic.io/docs/code-native-integrations/#adding-config-vars.
 * @param definition A Connection Config Var type object.
 * @returns This function returns a connection config var object that has the shape the Prismatic API expects.
 */
const connectionConfigVar = (definition) => definition;
exports.connectionConfigVar = connectionConfigVar;
/**
 * @param definition A Component Manifest type object.
 * @returns This function returns a component manifest object that has the shape the Prismatic API expects.
 */
const componentManifest = (definition) => definition;
exports.componentManifest = componentManifest;
/**
 * This function creates a component object that can be
 * imported into the Prismatic API. For information on using
 * this function to write custom components, see
 * https://prismatic.io/docs/custom-components/writing-custom-components/#exporting-a-component.
 * @param definition A ComponentDefinition type object, including display information, unique key, and a set of actions the component implements.
 * @returns This function returns a component object that has the shape the Prismatic API expects.
 */
const component = (definition) => (0, convertComponent_1.convertComponent)(definition);
exports.component = component;
/**
 * This function creates an action object that can be referenced
 * by a custom component. It helps ensure that the shape of the
 * action object conforms to what the Prismatic API expects.
 * For information on writing custom component actions, see
 * https://prismatic.io/docs/custom-components/writing-custom-components/#writing-actions.
 * @param definition An ActionDefinition type object that includes UI display information, a function to perform when the action is invoked, and a an object containing inputs for the perform function.
 * @returns This function validates the shape of the `definition` object provided, and returns the same action object.
 */
const action = (definition) => definition;
exports.action = action;
/**
 * This function creates a trigger object that can be referenced
 * by a custom component. It helps ensure that the shape of the
 * trigger object conforms to what the Prismatic API expects.
 * For information on writing custom component triggers, see
 * https://prismatic.io/docs/custom-components/writing-custom-components/#writing-triggers.
 * @param definition A TriggerDefinition type object that includes UI display information, a function to perform when the trigger is invoked, and a an object containing inputs for the perform function.
 * @returns This function validates the shape of the `definition` object provided, and returns the same trigger object.
 */
const trigger = (definition) => definition;
exports.trigger = trigger;
/**
 * This function creates a data source object that can be referenced
 * by a custom component. It helps ensure that the shape of the
 * data source object conforms to what the Prismatic API expects.
 * For information on writing custom component data sources, see
 * https://prismatic.io/docs/custom-components/writing-custom-components/#writing-data-sources.
 * @param definition A DataSourceDefinition type object that includes UI display information, a function to perform when the data source is invoked, and a an object containing inputs for the perform function.
 * @returns This function validates the shape of the `definition` object provided, and returns the same data source object.
 */
const dataSource = (definition) => definition;
exports.dataSource = dataSource;
/**
 * For information and examples on how to write inputs
 * for custom component actions and triggers, see
 * https://prismatic.io/docs/custom-components/writing-custom-components/#adding-inputs.
 * @param definition An InputFieldDefinition object that describes the type of an input for a custom component action or trigger, and information on how it should be displayed in the Prismatic WebApp.
 * @returns This function validates the shape of the `definition` object provided, and returns the same input object.
 */
const input = (definition) => definition;
exports.input = input;
/**
 * For information on writing custom component connections, see
 * https://prismatic.io/docs/custom-components/writing-custom-components/#adding-connections.
 * @param definition A DefaultConnectionDefinition object that describes the type of a connection for a custom component action or trigger, and information on how it should be displayed in the Prismatic WebApp.
 * @returns This functions validates the shape of the `definition` object provided and returns the same connection object.
 */
const connection = (definition) => definition;
exports.connection = connection;
/**
 * For information on writing custom component connections using on-premise resources, see
 * https://prismatic.io/docs/custom-components/writing-custom-components/#adding-connections.
 * @param definition An OnPremConnectionDefinition object that describes the type of a connection for a custom component action or trigger, and information on how it should be displayed in the Prismatic WebApp.
 * @returns This function validates the shape of the `definition` object provided and returns the same connection object.
 */
const onPremConnection = (definition) => definition;
exports.onPremConnection = onPremConnection;
/**
 * For information on writing custom component connections, see
 * https://prismatic.io/docs/custom-components/writing-custom-components/#adding-connections.
 * @param definition An OAuth2ConnectionDefinition object that describes the type of a connection for a custom component action or trigger, and information on how it should be displayed in the Prismatic WebApp.
 * @returns This functions validates the shape of the `definition` object provided and returns the same connection object.
 */
const oauth2Connection = (definition) => definition;
exports.oauth2Connection = oauth2Connection;
const componentManifests = (definition) => definition;
exports.componentManifests = componentManifests;
var util_1 = require("./util");
Object.defineProperty(exports, "util", { enumerable: true, get: function () { return __importDefault(util_1).default; } });
__exportStar(require("./types"), exports);
var testing_1 = require("./testing");
Object.defineProperty(exports, "testing", { enumerable: true, get: function () { return __importDefault(testing_1).default; } });
__exportStar(require("./errors"), exports);
