// .ignored/@prismatic-io/spectral/dist/types/IntegrationDefinition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isComponentReference = exports.isConnectionReferenceConfigVar = exports.isConnectionDefinitionConfigVar = exports.isDataSourceReferenceConfigVar = exports.isDataSourceDefinitionConfigVar = exports.isScheduleConfigVar = exports.isCodeConfigVar = void 0;
const isCodeConfigVar = (cv) => "dataType" in cv && cv.dataType === "code";
exports.isCodeConfigVar = isCodeConfigVar;
const isScheduleConfigVar = (cv) => "dataType" in cv && cv.dataType === "schedule";
exports.isScheduleConfigVar = isScheduleConfigVar;
const isDataSourceDefinitionConfigVar = (cv) => "dataSourceType" in cv && "perform" in cv && typeof cv.perform === "function";
exports.isDataSourceDefinitionConfigVar = isDataSourceDefinitionConfigVar;
const isDataSourceReferenceConfigVar = (cv) => "dataSourceType" in cv &&
    "dataSource" in cv &&
    (0, exports.isComponentReference)(cv.dataSource);
exports.isDataSourceReferenceConfigVar = isDataSourceReferenceConfigVar;
const isConnectionDefinitionConfigVar = (cv) => "inputs" in cv;
exports.isConnectionDefinitionConfigVar = isConnectionDefinitionConfigVar;
const isConnectionReferenceConfigVar = (cv) => "connection" in cv && (0, exports.isComponentReference)(cv.connection);
exports.isConnectionReferenceConfigVar = isConnectionReferenceConfigVar;
const isComponentReference = (ref) => typeof ref === "object" && ref !== null && "key" in ref && "component" in ref;
exports.isComponentReference = isComponentReference;
